query.cc

```diff
@@ -60,6 +60,20 @@ std::ostream& operator<<(std::ostream& out, const read_command& r) {
     return out;
 }
 
+lw_shared_ptr<query::read_command> reversed(lw_shared_ptr<query::read_command>&& cmd)
+{
+    return ::make_lw_shared(reversed(std::move(*cmd)));
```

Comment:
This is destroying *cmd, which might still be used by the caller. It really should have been lw_shared_ptr<const read_command>, but isn't (for historical reasons).